linked_list/linked_list.py

```diff
@@ -1,102 +1,183 @@
 
 # Defines a node in the singly linked list
 class Node:
-
-    def __init__(self, value, next_node = None):
+    def __init__(self, value, next_node=None):
         self.value = value
         self.next = next_node
 
 # Defines the singly linked list
 class LinkedList:
-    def initialize(self):
-      self.head = None # keep the head private. Not accessible outside this class
+    # def initialize(self):
+    def __init__(self):
+        self.head = None  # keep the head private. Not accessible outside this class
 
     # returns the value in the first node
     # returns None if the list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(n)
     def get_first(self):
-        pass
+        if self.head == None:
+            return None
 
+        return self.head.value
 
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(n)
+
     def add_first(self, value):
-        pass
+        new_node = Node(value)
+        new_node.next = self.head
+        self.head = new_node
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def search(self, value):

```

Comment:
👍 , but how is the space complexity O(n)?  What set of nodes are you creating?

Reply:
That's a good point. I was thinking about linked lists as a whole and not the function itself. I will go back and review them individually again. Thank you!